New backend: Infomaniak kDrive

Tofee opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?
Propose a backend implementation for Infomaniak's kDrive storage
<!--
Describe the changes here
-->
A new "kdrive" backend is added. The code was initially copied from the "pcloud" backend, and then modified to match the kDrive API.

#### Was the change discussed in an issue or in the forum before?
A request for this backend appeared some time ago, here: https://github.com/rclone/rclone/issues/8456
<!--
Link issues and relevant forum posts here.
-->

This is my first contribution to rclone, and my first attempt at the Go language, so while I'm trying to follow the good practices, be aware there might be many mistakes in this new backend... This is still a WIP, and this PR is intended to start the convergence on the code of the backend. 

The documentation for the backend is still missing; I'll add it soon.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

Tofee committed:
kdrive: initial commit for new backend

Signed-off-by: Christophe Chapuis <chris.chapuis@gmail.com>

Tofee committed:
kdrive: Filename encoding fixes

Tofee committed:
kdrive: fix some usecases

Tofee committed:
kdrive: remove unsupported OpenWriterAt

Tofee committed:
kdrive: disable broken features and fix OpenRange

Tofee committed:
kdrive: all backend tests are now OK

Tofee committed:
kdrive: cleanup

Tofee committed:
kdrive: add backend to testall config.yaml

Tofee committed:
kdrive: add documentation

Signed-off-by: Christophe Chapuis <chris.chapuis@gmail.com>

Tofee committed:
kdrive: cleanup unused oauth

Tofee committed:
kdrive: fix remaining tests

Tofee committed:
kdrive: fix lint errors

Tofee committed:
kdrive: fix path of items when using root id folder

Tofee committed:
kdrive: fix hash update after content change

Gurvan89 reviewed (commented):
I added some comment about the docs.

I think you should also add the link of the documentation in the main readme file.

Tofee committed:
kdrive: docs: fix incorrect names

There were pending references to "pcloud", it should be "kdrive".

Tofee commented:
> I think you should also add the link of the documentation in the main readme file.

There is already a link to the documentation in the main readme file, it's handled via bin/make_manual.py and docs/content/_index.md.

Tofee committed:
kdrive: add reference in main README file

Tofee commented:
> Hello 👋 ,

Hi!

> Thank you very much for your upstream contribution and for the time and effort you put into it. We really appreciate it!

Thanks !

> To give you some context: at Infomaniak, we’ve recently decided to officially develop our own kDrive backend implementation internally. We’re working on a different implementation with slightly different constraints and architectural choices (for example, using the listing/full route instead of a recursive approach).

Nice to hear you've also begun working on a backend; do you plan to release something, or is it going to stay as an internal tool?

> That said, we do not plan to close your PR. There are definitely interesting and valuable ideas in your contribution, and we’ll take the time to review it properly and see how we can adapt or build upon it within our current direction.

In the end, it would be preferable to have one converged backend implementation, whatever shape it takes. I'll leave this PR continue its way forward, see when/if it gets merged, and then we can see how to improve it.

> Thanks again for your involvement and for contributing upstream. It’s genuinely appreciated.

Thanks again, it's also nice to hear from you!

matthieu-mabillard commented:
Hello 👋 ,

Sorry if my previous message wasn't 100% clear! To give you some context, the reason I reached out was precisely to coordinate our efforts: we are currently finalizing our own native kDrive backend for rclone and were planning to open a PR very soon on this repo.

Since we are working toward the same goal, I thought it would be best to align before anything is merged. My goal is to ensure we can review and converge our two implementations rather than having fragmented versions 😄 

By combining our work, we can offer the most robust and optimal experience for kDrive users. 

Looking forward to collaborating on this!

mallsbill committed:
kdrive: add multipart upload and internal tests

mallsbill committed:
kdrive: improve error management

mallsbill committed:
kdrive: bearer token instead of oauth2

mallsbill committed:
kdrive: fix PublicLink

ikbuben committed:
kdrive: remove unused Profile

ikbuben committed:
kdrive: remove some unused types

ikbuben committed:
kdrive: consistent ResultStatus receiver type

ikbuben committed:
kdrive: quick cleanup

mallsbill committed:
kdrive: add ChunkSize option to multipart and refuse files with no size

mallsbill committed:
kdrive: fix hash control on large file upload

mallsbill committed:
kdrive: use get file by path api

mallsbill committed:
kdrive: SetModTime implementation

mallsbill committed:
kdrive: fix options and code fixme

mallsbill committed:
kdrive: fix type cancel session + change path findItemInDir

mallsbill committed:
kdrive: Add item hasChildren on listing

mallsbill committed:
kdrive: fix directory in double en walk.GetAll

mallsbill committed:
kdrive: fix directory in double en walk.GetAll

mallsbill committed:
kdrive: use private folder as default root

mallsbill committed:
kdrive: delete OpenChunkWriter implementation because kdrive can't manage file without a size

mallsbill committed:
kdrive: optimize update and reorder code

mallsbill committed:
kdrive: remove ignore FsOpenChunkWriter test

mallsbill committed:
kdrive: fix options

ikbuben committed:
kdrive: doc cleanup

mallsbill committed:
kdrive: fix conflicts

ikbuben committed:
kdrive: fix root_folder_id option

ikbuben committed:
kdrive: fix doc indent

mallsbill committed:
kdrive: optimize api call on init with context cache

mallsbill committed:
kdrive: cache not found on init

mallsbill committed:
kdrive: fix Public link unlink and expiration

mallsbill committed:
kdrive: fix CleanUp

mallsbill committed:
kdrive: use depth unlimited for ListR

mallsbill committed:
kdrive: Public Link, update expiration date if provided

mallsbill committed:
kdrive: Add cache on not found and clear logs

Tofee commented:
> @Tofee We opened a PR toward this one here: [Tofee#1](https://github.com/Tofee/rclone/pull/1), if you can review it 😄 . It contains the changes we have made based on your initial work. Thanks for your collaboration!

Thanks @matthieu-mabillard, it's a very nice set of improvements! I'll have a look this week, and get back to you. 
From a quick look, it looks like I'll need to recreate a new config entry, right?

mallsbill committed:
kdrive: Delete test files folder

mallsbill committed:
kdrive: treating forbidden on expiration date as success for PublicLink test

mallsbill committed:
kdrive: disable cache and fix encoding

mallsbill committed:
kdrive: fix copy should replace existing file

mallsbill committed:
kdrive: ignoretests cmd/bisync and cmd/gitannex

mallsbill committed:
kdrive: manage conlict on update by creating a new version of a file and fix maxChunkSize

matthieu-mabillard commented:
@ncw Is there anything specific blocking this from being merged? Happy to address whatever's needed. Thanks!

GabriF commented:
Hello, I have tried to use kdrive backend cloning from tofee/rclone because I had to migrate from google drive to kdrive. It took me about an hour to complete the configuration, I don't understand why the configuration wizard asked me to either generate or type in my own password. I eventually figured out to create a token from https://manager.infomaniak.com/v3/XXX/ng/profile/user/token/list and then copy it, finally it worked. The doc tells that rclone should guide me through the browser but that doesn't happen.

JameDevOfficial commented:
> Hello, I have tried to use kdrive backend cloning from tofee/rclone because I had to migrate from google drive to kdrive. It took me about an hour to complete the configuration, I don't understand why the configuration wizard asked me to either generate or type in my own password. I eventually figured out to create a token from https://manager.infomaniak.com/v3/XXX/ng/profile/user/token/list and then copy it, finally it worked. The doc tells that rclone should guide me through the browser but that doesn't happen.

Ye, I had the same issue, but eventually figured it out.

Tofee commented:
@ncw sorry to ping you again, but this backend had been ready for merge for quite some time now... Is there something missing on our side?

ncw commented:
Sorry all been a bit overloaded! Hopefully we can get this merged for v1.75 @Tofee @matthieu-mabillard 

Reading @GabriF and @JameDevOfficial comments it looks like the config UI isn't quite right?

I see all the tests are passing which is great :-)

Tofee commented:
> Reading @GabriF and @JameDevOfficial comments it looks like the config UI isn't quite right?
>

Well yes the configuration might be a little confusing here: the access token is generated on the Infomaniak account page, it's not an OAuth process like several backend use. I guess the wizard always expects the token to be generated through OAuth, thus the misunderstanding here with the users...
Not sure how we could avoid this default wizard message here?

I'll have a look at the lint/AIX errors too.

Tofee committed:
kdrive: fix lint errors

Tofee committed:
kdrive: fix build on Plan9

mabac commented:
Hey Heroes! I'm using this since a week or so and have no issues. If you need any extra testing or eyes on something to get it landed, I'm game. :)

ncw commented:
@Tofee Looks like you need to fix the merge conflicts. Let me know if you want me to do it for you (it is a bit of a black art!)

Tofee commented:
> @Tofee Looks like you need to fix the merge conflicts. Let me know if you want me to do it for you (it is a bit of a black art!)

I think I got the merge right, hopefully there was just one conflict in a yaml file, so it's fairly easy.
I'm running the "build" workflow on my branch now to check if I didn't miss something, let's see..

Tofee commented:
@ncw ok, all good on my side, the build workflow passes and the tests are still 100% OK.

binomnom commented:
First of all, thank you guys so much for putting in all this work. We have just had to migrate data into kDrive, and this has made it so much easier.
Thanks especially to @Tofee for picking up the torch and @matthieu-mabillard & @mallsbill for being good sports and contributing!

## With that being said, I encountered one issue when using the backend I believe should be addressed still:

When running a sync on two identical files, instead of updating a changed modification time, it **reuploads** the file with the updated modTime. Not efficient, as it causes a huge amount of additional traffic.
But the real problem is that kDrive has a **versioning feature** that transparently uses the purchased storage space. This means every time rclone needs to update modTime, a new "version" is created, doubling the storage need for the file.
I can't think of a case where someone would want hash-identical files with differing modTimes to be versioned.
In our case, **500GB of files turned into 680GB of storage footprint** on kDrive (after running partial uploads using different uplinks and then running one cleanup pass over everything).

### Example:
```
2026/05/15 08:27:25 DEBUG : rclone: Version "v1.75.0-beta.9783.c3ae9d395.tofee/kdrive" starting with parameters ["./rclone" "sync" "--modify-window" "2s" "/run/media/<redacted>/<file>" "kdrive:<redacted>/" "-vv"]
...
2026/05/15 08:27:25 INFO  : directoryID <redacted> leaf <redacted>
2026/05/15 08:27:26 DEBUG : <redacted>: size = 8933376 OK
2026/05/15 08:27:26 DEBUG : <redacted>: Modification times differ by 21811h37m52s: 2023-11-16 11:20:58 +0100 CET, 2026-05-13 07:58:50 +0200 CEST
2026/05/15 08:27:26 DEBUG : <redacted>: xxh3 = ba688a6fbbd79bb0 OK
2026/05/15 08:27:27 NOTICE: kdrive root '<redacted>': Forced to upload files to set modification times on this backend.
2026/05/15 08:27:27 INFO  : <redacted>: src and dst identical but can't set mod time without re-uploading
2026/05/15 08:27:27 DEBUG : <redacted>: size = 8933376 OK
2026/05/15 08:27:27 DEBUG : <redacted>: xxh3 = ba688a6fbbd79bb0 OK
2026/05/15 08:27:27 INFO  : <redacted>: Copied (replaced existing)
```

### Possible fixes

I see in the API docs that there is an [endpoint for updating modification date](https://developer.infomaniak.com/docs/api/post/3/drive/%7Bdrive_id%7D/files/%7Bfile_id%7D/last-modified). This endpoint has a limitation of only being able to update the modTime to 24h from now. (This is not quite true, the API actually allows up to -5 days):
```curl
curl --location 'https://api.infomaniak.com/3/drive/<redacted>/files/<redacted>/last-modified' \
--header 'Authorization: Bearer <redacted>' \
--header 'Content-Type: application/json' \
--data '{
    "last_modified_at": 1511612678
}'
```

```json
{
    "result": "error",
    "error": {
        "code": "validation_failed",
        "description": "Validation failed",
        "errors": [
            {
                "code": "validation_rule_after",
                "description": "The last modified at must be a date after -5 days.",
                "context": {
                    "attribute": "last_modified_at"
                }
            }
        ]
    }
}
```
@matthieu-mabillard could this limit be lifted? I do see the logic behind it, but it would be awesome to be able to not have to reupload the whole file to keep a historic modTime.

To remedy, either;
1. Infomaniak lifts the limit on the `last-modified` endpoint and this endpoint is used in the backend (Preferred); or
2. As a band-aid solution, after uploading the file to set modTime, the `DELETE versions/{version_id}` endpoint is called to remove the version that has been replaced.

### Workaraound
For now, running the `--no-update-modtime` flag in your rclone command will prevent the additional traffic at the cost of inconsistent modification time metadata.



## Another small thing:
rclone regularly runs into the API rate limit, leading to `kDrive error: (429 Too Many Requests: 429 Too Many Requests)`


### Output of `rclone version`:
```bash
rclone v1.75.0-beta.9783.c3ae9d395.tofee/kdrive
- os/version: arch (64 bit)
- os/kernel: 7.0.3-arch1-2 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.26.2-X:nodwarf5
- go/linking: dynamic
- go/tags: none
```

mallsbill committed:
kdrive: fix set last_modified_at on multipart upload

mallsbill committed:
kdrive: add oauth2 authentification with PKCE

Tofee commented:
@matthieu-mabillard I discussed with a user that had difficulties accessing the "Common documents" and "Shared" folders, and although it's something that can be configured in the Advanced Options, it's maybe too hidden currently.

What would you say about moving this root folder option into the regular options? On the pro side, it would make it obvious for "Organization" users how to access the shared folders, but on the "con" side, for a simple free account with only "Private" documents, it might be confusing. Not sure what to decide here.

mallsbill committed:
kdrive: add large files nested hash management

matthieu-mabillard commented:
@Tofee It's a good feedback ! We discussed it, and we think it is better for now to keep the option in the advanced settings. Most of our users will only use "Private documents," so we should simplify the behavior as much as possible.

matthieu-mabillard commented:
@binomnom Hello 👋 , thanks for the feedback!

Indeed, it was a problem, so we made two changes:
1. The rule has been fixed (we deleted it). It is in production now on the kDrive API
2. We opened a PR that awaits the review from @Tofee  (CF: https://github.com/Tofee/rclone/pull/4) to fix the `last_modified_at` on multiple part upload

mallsbill committed:
kdrive: add oauth2 authentification with PKCE

- revert changes linting

mallsbill committed:
kdrive: add large files nested hash management

- fix linting

mallsbill committed:
kdrive: fix alignment multipart sessionReq

Tofee committed:
kdrive: fix oauthutil build

binomnom commented:
@Tofee I have encountered a new issue:
**The new hash function for multipart uploads seems to be broken:**
- single-part uploads work fine: `./rclone copy small-file.txt kdrive-backend:`
- multi-part uploads get uploaded, then always fail the hash check and get deleted again:
   ```
   ./rclone copy big-file.bin kdrive-backend:
   rclone: corrupted on transfer: kdrivehash hashes differ src(Local file system at /home/<redacted>/big-file.bin) "6dc7f3ca4270f4a6" vs dst(kdrive root '') "241a2d1410e42244"
   ```

And a second issue, though this may be user error:  
**If I set `root_folder_id` to "shared", I can't list or copy files at all.**

I understand that setting `root_folder_id` to "shared" should set the remote's root to the `/shared-with-me/` directory as seen in the web UI, correct?
My situation is as follows: I have one folder called "Backups" that someone shared with me. I'm trying to copy to this folder using `kdrive-backend:Backups/` while the `root_folder_id` is set to "shared".

Is that the correct approach? I have also tried setting it to `4` instead of `shared`, with the same result. If I set it to the ID of the "Backups" folder directly, it works.

Output of `./rclone version`:
```
rclone v1.75.0-DEV
- os/version: debian 13.5 (64 bit)
- os/kernel: 6.12.86+deb13-amd64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.25.0
- go/linking: dynamic
- go/tags: none

mallsbill committed:
kdrive: fix hash check

Tofee commented:
@binomnom I could reproduce the issue with the "shared" root folder. Somehow, I couldn't list the files there either.
I'm really not sure what the "Shared" folder is for, as there is a dedicated API for "shared_with_me" files and for "my_shared" files...

For the hash issue, I'll let @matthieu-mabillard have a look, he introduced it last week so he knows better

mallsbill commented:
@Tofee I created a PR for the hash issue and I'm going to look for shared files.

mallsbill committed:
kdrive: fix hash check

binomnom commented:
@mallsbill thank you, I can confirm hashing now works on my side, which is great news! I also verified that updateModtime works now as well as I haven't tried that since it was implemented. This is getting exciting!

Are you still looking into the "shared-with-me" issue? If so, here's what I have been able to figure out:

if I use `--dump requests,responses`, I can see that the reply from the API actually works correctly, both with `shared`, and with `4` as `root_folder_id`. Somehow, rclone silently fails to parse the directory structure from this and just returns when using any listing comand (`ls`, `lsd`, ...).
When using `sync` or `copy`, the GET request for the specified directory gets a 404 response from the API. Unsure what format the API is expecting the paths to be, though.



### Example:  (The following output is very abridged):

**1: If it is set to `shared`:**
Request by rclone:
```
DEBUG : GET /3/drive/<redacted>/files/1/name?name=Shared&with=path%2Chash HTTP/1.1
```
Response by API:
```json
{
  "result": "success",
  "data": {
    "id": 4,
    "name": "share_files",
    "path": "/Shared/share_files",
    "type": "dir",
    "status": "ok",
    "visibility": "is_shared_space",
    "drive_id": <redacted>,
    "depth": 1,
    "created_by": null,
    "created_at": null,
    "added_at": <redacted>,
    "last_modified_at": <redacted>,
    "last_modified_by": null,
    "revised_at": <redacted>,
    "updated_at": <redacted>,
    "parent_id": 1,
    "shared_root_id": null,
    "color": null
  }
}
```

2: From here on, the requests are the same for both `shared` and `4`:

```
DEBUG : GET /3/drive/<redacted>/files/4?with=path HTTP/1.1
```
```json
{
  "result": "success",
  "data": [
    {
      "id": <redacted>,
      "name": "Test",
      "path": "/Shared/shared/FULL/PATH/OF/FOLDER/IN/ORIGINATING/DRIVE/Test",
      "type": "dir",
      "status": "ok",
      "visibility": "is_in_shared_space",
      "drive_id": <redacted>,
      "depth": 4,
      "created_by": <redacted>,
      "created_at": null,
      "added_at": <redacted>,
      "last_modified_at": <redacted>,
      "last_modified_by": <redacted>,
      "revised_at": <redacted>,
      "updated_at": <redacted>,
      "parent_id": 1,
      "shared_root_id": 4,
      "color": null
    }
  ],
  "response_at": 1781072303,
  "cursor": null,
  "has_more": false
}
```

3: When using `copy file.ext kdrive-remote:Test` it goes straight from step 1 to:
```
DEBUG : GET /3/drive/<redacted>/files/4/name?name=Test&with=path%2Chash HTTP/1.1
```
```json
{
  "result": "error",
  "error": {
    "code": "object_not_found",
    "description": "Object not found",
    "context": {
      "model": "GenericFile"
    }
  }
}
```



Config contents:
```
[kdrive-remote]
type = kdrive
drive_id = XXX
root_folder_id = shared        # or 4
token = XXX
```

mallsbill commented:
Hi @binomnom , I can resolved the listing by using the api route dedicated to "shared-with-me" virtual folder, but I still have a problem to manage download.
Rclone search an item by its name (leaf) but technically you can have multiple item with the same name in the shared with me folder (from different drive for example).

I don't currently see how to handle this properly, because this folder is special and doesn't work like a normal filesystem.

Tofee commented:
@mallsbill in this case, wouldn't it make sense to remove the "shared" root folder id, until it is properly handled by the backend?

mallsbill commented:
@Tofee Yes, that would be the best thing to do.

Tofee committed:
kdrive: remove unsupported shared root folder id

Tofee commented:
I think we're in a quite good state now.

@ncw I re-ran the build tests, backend tests and integration tests, everything is fine. This might be a good time to merge this (long) PR ?

Edit: after a rebase, the backend/cache linux test is failing. I guess it's the same for upstream.
Edit2: a [second run](https://github.com/Tofee/rclone/actions/runs/27371901287) of the same build workflow passed fine, looks like an instability in the backend/cache tests

richerve commented:
@Gurvan89 is there anything else required here?

Tofee commented:
@matthieu-mabillard do we also need to apply this kind of fix https://github.com/rclone/rclone/commit/5d4c2c32a09f2f96af732e540891a7a48c1c74e1 ? I'm not sure what modtime precision is used on server side for kDrive.

binomnom commented:
> @matthieu-mabillard do we also need to apply this kind of fix [5d4c2c3](https://github.com/rclone/rclone/commit/5d4c2c32a09f2f96af732e540891a7a48c1c74e1) ? I'm not sure what modtime precision is used on server side for kDrive.

I don't know about the server side, but a query of the kDrive API always returns second precision: https://developer.infomaniak.com/docs/api/get/3/drive/%7Bdrive_id%7D/files/%7Bfile_id%7D

So it's probably a good call, though I haven't tested it yet.

Tofee committed:
kdrive: fix incorrect modtime after setting a file's modtime

SetModTime kept the caller's full precision modtime in memory while
the server stores second precision, so until the object was re-read
the modtime did not match what a fresh listing would report. This
broke wrappers which compare exact modtimes, such as the hasher
backend's fingerprint and the VFS cache. Truncate the modtime to
second precision to match what the server stores.

Tofee commented:
OK I've incldued the modTime fix, and rebased on upstream's latest release. We're still 100% OK on the tests:
<img width="500" alt="image" src="https://github.com/user-attachments/assets/a16cac7f-4f42-4e4a-86dc-86c27f6a0c60" />

@ncw Just after the release is out, this is a finally good time for a merge, don't you think?

JorisBodin commented:
@ncw Do you think a merge is possible for the upcoming release? 😊

nuxlight commented:
I have no see something wrong on Go code, and I have tested with my Infomaniak account that's work 🎉 

I have just to add the `docs/data/backends/kdrive.yaml` file to remove this error after local build : 
````
2026/08/22 09:10:48 ERROR : internal error: no overview data found for "kdrive"
````

Thanks for the work 🙏

Tofee commented:
> I have just to add the `docs/data/backends/kdrive.yaml` file to remove this error after local build :

@nuxlight Not sure I understand this comment correctly, is there a file missing that would trigger a build error in some case?

nuxlight commented:
> > I have just to add the `docs/data/backends/kdrive.yaml` file to remove this error after local build :
> 
> @nuxlight Not sure I understand this comment correctly, is there a file missing that would trigger a build error in some case?

Sorry if I don’t explain this very well 🙏
When I want to test this PR on my Mac, I need to create the file docs/data/backends/kdrive.yaml. I quickly generated it with an LLM:

```yaml
backend: kdrive
name: kdrive
tier: Tier 2
maintainers: Community
features_score: 5
integration_tests: Unknown
data_integrity: Unknown
performance: Unknown
adoption: Rare
docs: Basic
security: Unknown
virtual: false
remote: 'TestKDrives:'
features:
- About
- Copy
- Move
hashes:
- kdrivehash
precision: 1000000
```
